src/goto-instrument/cover_basic_blocks.cpp

```diff
@@ -0,0 +1,227 @@
+/*******************************************************************\
+
+Module: Coverage Instrumentation
+
+Author: Daniel Kroening
+
+\*******************************************************************/
+
+/// \file
+/// Basic blocks detection for Coverage Instrumentation
+
+#include "cover_basic_blocks.h"
+
+#include <util/format_number_range.h>
+#include <util/message.h>
+#include <util/string2int.h>
+
+cover_basic_blockst::cover_basic_blockst(const goto_programt &_goto_program)
+{
+  bool next_is_target = true;
+  unsigned current_block = 0;
+
+  forall_goto_program_instructions(it, _goto_program)
+  {
+    // Is it a potential beginning of a block?
+    if(next_is_target || it->is_target())
+    {
+      // We keep the block number if this potential block
+      // is a continuation of a previous block through
+      // unconditional forward gotos; otherwise we increase the
+      // block number.
+      bool increase_block_nr = true;
+      if(it->incoming_edges.size() == 1)
+      {
+        goto_programt::targett in_t = *it->incoming_edges.begin();
+        if(
+          in_t->is_goto() && !in_t->is_backwards_goto() &&
+          in_t->guard.is_true())
+        {
+          current_block = block_map[in_t];
+          increase_block_nr = false;
+        }
+      }
+      if(increase_block_nr)
+      {
+        block_infos.push_back(block_infot());
+        block_infos.back().representative_inst = it;
+        block_infos.back().source_location = source_locationt::nil();
+        current_block = block_infos.size() - 1;
+      }
+    }
+
+    INVARIANT(
+      current_block < block_infos.size(), "current block number out of range");
+    block_infot &block_info = block_infos.at(current_block);
+
+    block_map[it] = current_block;
+
+    // update lines belonging to block
+    const irep_idt &line = it->source_location.get_line();
+    if(!line.empty())
+      block_info.lines.insert(unsafe_string2unsigned(id2string(line)));
+
+    // set representative program location to instrument
+    if(
+      !it->source_location.is_nil() &&
+      !it->source_location.get_file().empty() &&
+      !it->source_location.get_line().empty() &&
+      block_info.source_location.is_nil())
+    {
+      block_info.representative_inst = it; // update
+      block_info.source_location = it->source_location;
+    }
+
+    next_is_target =
+#if 0
+      // Disabled for being too messy
+      it->is_goto() || it->is_function_call() || it->is_assume();
+#else
+      it->is_goto() || it->is_function_call();
+#endif
+  }
+
+  for(auto &block_info : block_infos)
+    update_covered_lines(block_info);
+}
+
+unsigned cover_basic_blockst::block_of(goto_programt::const_targett t) const
+{
+  block_mapt::const_iterator it = block_map.find(t);
+  INVARIANT(it != block_map.end(), "instruction must be part of a block");
+  return it->second;
+}
```

Comment:
I'm kind of surprised we don't have a utility function for 'find basic blocks', similarly to finding loops.  It would be good if this could be split out to be something general.

Reply:
#1686